Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean

```diff
@@ -216,18 +216,22 @@ theorem affineLocally_iff_affineOpens_le (hP : RingHom.RespectsIso @P) {X Y : Sc
       ext1; refine' Set.image_preimage_eq_inter_range.trans (Set.inter_eq_left_iff_subset.mpr _)
       erw [Subtype.range_val]
       convert e
-    have := H ⟨(Opens.map (X.ofRestrict U'.openEmbedding).1.base).obj V.1, ?_⟩
+    have := H ⟨(Opens.map (X.ofRestrict U'.openEmbedding).1.base).obj V.1, ?i⟩
     erw [← X.presheaf.map_comp] at this
-    rw [← hP.cancel_right_isIso _ (X.presheaf.map (eqToHom _)), Category.assoc, ←
-      X.presheaf.map_comp]
-    convert this using 1
-    · -- Porting note: makes instance metavariable like in Lean 3
+    /- Note: `?i` (formerly unnamed) was previously returned by `convert`, which produced the
+    following porting note:
+    -- Porting note: makes instance metavariable like in Lean 3
+    -/
+    case i =>
       apply
         (@isAffineOpen_iff_of_isOpenImmersion _ _ (@Scheme.ofRestrict _ X U'.inclusion _) ?_ _).mp
       -- Porting note: was convert V.2
       erw [e']
       apply V.2
       infer_instance
+    rw [← hP.cancel_right_isIso _ (X.presheaf.map (eqToHom _)), Category.assoc, ←
+      X.presheaf.map_comp]
+    convert this using 1
```

Comment:
I remember struggling with this `convert` [here](https://github.com/leanprover-community/mathlib4/commit/90689c421c95e1f3265daf4dd28958456583a754#diff-adf113095495d457bf63a8988b82ce79bb13ae125a8d285d533c039786e196b6R224-R232). The spooky action at a distance was maddening, and it's nice to know that there's an answer now to what was going on. The porting note might not make any sense given that no one really investigated where the metavariable was coming from. (Do you know? Was simply from the type of `this`?)

Could you reorganize the proof a little? I think we want the `rw`, `convert`, and `dsimp` line to come before the `case i`, and maybe remove the `·` from the `dsimp` line? Then the `case i` block can come after this.

Would you mind checking the "Porting note: was convert V.2" too? There's a chance `convert` works now. If not, don't worry about it. Thanks.

Reply:
Nice! Hmm...`?i : (Opens.map (Scheme.ofRestrict X (_ : OpenEmbedding ↑(Opens.inclusion U'))).val.base).obj ↑V ∈ Scheme.affineOpens (Scheme.restrict X (_ : OpenEmbedding ↑(Opens.inclusion ((Opens.map f.val.base).obj ↑U))))` is in the definition of `this` created by `have`, which then tunneled into the goal state due to `convert`. And then I suppose `apply (@isAffineOpen_iff_of_isOpenImmersion _ _ (@Scheme.ofRestrict _ X U'.inclusion _) ?_ _).mp` introduced the actual instance argument somehow—the thing before the `mp` has type
```lean
IsAffineOpen
    ((PresheafedSpace.IsOpenImmersion.openFunctor ?m.95244).obj
      ((Opens.map (Scheme.ofRestrict X (_ : OpenEmbedding ↑(Opens.inclusion U'))).val.base).obj ↑V)) ↔
  IsAffineOpen ((Opens.map (Scheme.ofRestrict X (_ : OpenEmbedding ↑(Opens.inclusion U'))).val.base).obj ↑V)
```
and we have `?m.95244 : IsOpenImmersion (Scheme.ofRestrict X (_ : OpenEmbedding ↑(Opens.inclusion ((Opens.map f.val.base).obj ↑U))))`. `infer_instance` closes this goal (`IsOpenImmersion (Scheme.ofRestrict X (_ : OpenEmbedding ↑(Opens.inclusion ((Opens.map f.val.base).obj ↑U))))`, just to check it's the same.

Unfortunately it looks like `convert V.2` doesn't work still; I looked at the mathported source, and the `infer_instance` complains that it was expecting a typeclass instance. (Instead, there are two goals: `IsOpenAffine` and another more complicated one.)

I was able to reorgnize the proof and eliminate `case i` altogether—it just naturally comes to the top of the goal list. Is that good proactice? Or is `case i` more readable?

Reply:
How about you name the metavariable and wrap the `apply ... infer_instance` part back up in a `case`? I think that's more maintainable, since then someone can come by later and better see where the start/end of each goal is.

If you [look at the mathlib3](https://github.com/leanprover-community/mathlib/blob/442a83d738cb208d3600056c489be16900ba701d/src/algebraic_geometry/morphisms/ring_hom_properties.lean#L192-L201) version, your fixes have recovered the original order of the goals. There's a difference in the new `convert` and now the `{ apply_instance }` isn't needed. That means we don't need to wrap the `dsimp ... apply` part of the proof in a block anymore since `convert` is only generating a single goal. By the way that `apply e'.symm` seems like it's meant to be `exact e'.symm` judging by the Lean 3 source.

Once you do that, I think you can safely delete your note. I don't think you'll need to leave a porting note for naming the metavariable either since the port is over.

Reply:
Great! I quite like `case <h> =>` syntax just for readability, but I don't see it often in mathlib, so I wondered if it was preferred only when necessary.

I rather unimaginatively named the hypothesis `h'` (as `h` was taken in the mvar context), since I'm not familiar with what this hypothesis is really saying. It seems to be saying that something is an affine open set, but I'm not willing to commit to that. If you or someone wants to suggest a more meaningful name, though, feel free to let me know or add a commit. :)

Reply:
I think you don't see `case` much for the simple reason that Lean 3 didn't have `?h` metavariables! We only had `_`, and tactics frequently mangled the auto-generated tags.

Re `h'`, it's better than `_`, so I think we can resolve this conversation. I don't think any future reviewer will suggest you change it (unless you keep this conversation unresolved!).

Reply:
> Lean 3 didn't have ?h metavariables!

Ah, that would do it! :D